src/stan-users-guide/parallelization.Rmd

```diff
@@ -1,12 +1,314 @@
-# Map-Reduce  {#map-reduce.chapter}
+# Parallelization  {#parallelization.chapter}
+
+Stan has two mechanisms for parallelizing calculations used in a
+model: reduce with summation and rectangular map.
+
+The advantages of reduce with summation are:
+
+1. More flexible argument interface, avoiding the packing and
+   unpacking that is necessary with rectanguar map.
+2. Partitions data for parallelization automatically (this is done manually in rectanguar map).
+3. Is easier to use.
+
+The advantages of rectangular map are:
+
+1. Returns a list of vectors, while the reduce summation returns only a scalar.
+2. Can be parallelized across multiple cores and multiple
+   computers, while reduce summation can only parallelized across multiple
+   cores on a single machine.
+
+The actual speedup gained from using these functions will depend on
+many details. It is strongly recommended to only parallelize the
+computationally most expensive operations in a Stan
+program. Oftentimes this is the evaluation of the log likelihood for
+the observed data. Since only portions of a Stan program will run in
+parallel, the maximal speedup one can achieve is capped, a phenomen
+described by [Amdahl's
+law](https://en.wikipedia.org/wiki/Amdahl's_law).
+
+## Reduce-Sum { #reduce-sum }
+
+The higher-order reduce with summation facility maps evaluation of a
+function `g: U -> real`, which returns a scalar, to a list of type
+`U[]`, `{ x1, x2, ... }`, and performs as reduction operation a sum
+over the results. For instance, for a sequence of ```x``` values of
+type ```U```, ```{ x1, x2, ... }```, we might compute the sum:
+
+```g(x1) + g(x2) + ...```
+
+In probabilistic modeling this comes up when there are $N$
+conditionally independent terms in a likelihood. Because of the
+conditional independence, these terms can be computed in parallel. If
+dependencies exist between the terms, then this isn't possible. For
+instance, in evaluating the log density of a Gaussian process then
+summation of independent terms isn't applicable.
+
+For efficiency reasons the reduce function doesn’t work with the
+element-wise evaluated function `g`, but instead requires the partial
+sum function ```f: U[] -> real```, where ```f``` computes the partial
+sum corresponding to a slice of the sequence ```x``` passed in. Due to the
+the associativity of the sum reduction it holds that:
+
+```
+g(x1) + g(x2) + g(x3) = f({ x1, x2, x3 })
+                      = f({ x1, x2 }) + f({ x3 })
+                      = f({ x1 }) + f({ x2, x3 })
+					  = f({ x1 }) + f({ x2 }) + f({ x3 })
+```
+
+With the partial sum function ```f: U[] -> real``` reduction of a
+large number of terms can be evaluated in parallel automatically, since the
+overall sum can be partitioned into arbitrary smaller partial
+sums. The exact partitioning into the partial sums is not under the
+control of the user. However, since the exact numerical result will
```

Comment:
I find this confusing - what do you mean by "exact partitioning"?  a partition includes consecutive terms in x, no?  but slicing means that not all consecutive terms are in the same partition?  grainsize fixes the start/end indices, i.e. the call graph?
not sure if other readers would be similarly confused.

Reply:
A partition includes consecutive terms in x - that's correct. The thing is that for `reduce_sum` the user does not have the control over what happens exactly. Take 

```
g(x1) + g(x2) + g(x3) = f({ x1, x2, x3 })
                      = f({ x1, x2 }) + f({ x3 })
                      = f({ x1 }) + f({ x2, x3 })
					  = f({ x1 }) + f({ x2 }) + f({ x3 })
```

There are 4 ways to compute this sum and the user cannot control which one is executed. Only if you use the `reduce_sum_static` variant, then you can enforce always the same call graph - which one is determined by `grainsize`; but conceptually the user cannot - by construction - control which of the 4 variants are executed. So the "exact partitioning" is not under the users control.

If that was not clear... then feel free to suggest something more clear.